core/primitives/src/transaction.rs

```diff
@@ -415,6 +416,8 @@ pub struct ExecutionOutcome {
     pub receipt_ids: Vec<CryptoHash>,
     /// The amount of the gas burnt by the given transaction or receipt.
     pub gas_burnt: Gas,
+    /// The amount of compute time spent by the given transaction or receipt.
+    pub compute_usage: Compute,
```

Comment:
Thanks for spotting this, I was not aware of this consideration.

This makes me wonder why fields like `gas_burnt` need(?) to be a part of `ExecutionOutcome` and whether the same reason applies to `compute_usage`. I'll ask on Zulip about this.

> Hm, actually I was hoping we didn't need to include compute_usage as an explicit field on the outcome. Could we instead compute it on the fly?

Do you mean computing it based on the profile data in `metadata`? This is possible in some cases (host function calls) but breaks down in others that don't populate the profiling information (e.g. `apply_delegate_action`) and so would require more work

> Or maybe we can store the information in a different place? For example, it's somewhat hacky but if we add the field to the metadata we could just introduce ExecutionMetadata::V4. As we already have metadata versioned, it wouldn't break any existing proofs or DB serialized data.

This likely would be easier than the approach with profiles considered above, though I think it will change the code in roughly the same places, so maybe it's worth investing a bit more effort and properly computing profiles there.

For the approaches above that extend the `ExecutionOutcome.metadata`, are there any migration considerations I should be aware of?

Reply:
I've implemented an approach with profiles in https://github.com/near/nearcore/pull/8829, it's not that bad but feels pretty hacky due to abusing profiles. I'll see if I can improve it.

Reply:
> This makes me wonder why fields like gas_burnt need(?) to be a part of ExecutionOutcome and whether the same reason applies to compute_usage. I'll ask on Zulip about this.

Well, gas burnt was there before profiles were added. And profiles were initially only nice-to-have extra data.

> For the approaches above that extend the ExecutionOutcome.metadata, are there any migration considerations I should be aware of?

I think only one thing: You must add the new enum variant at the end, because borsh uses the variant index as defined by the order in code to (de-)serialize enums. Inside the enum variant, I think you can use any type you like. (See also the borsh spec on https://borsh.io/)

> I've implemented an approach with profiles in https://github.com/near/nearcore/pull/8829, it's not that bad but feels pretty hacky due to abusing profiles. I'll see if I can improve it.

Yeah it looks like it works but I agree regarding the profile abuse. I see two general directions for clean solutions here:
- Add the new field to execution outcome and do the necessary DB migration. And check if we need to do something else for proofs etc.
- Go all-in with profiles. I.e. make profiles tracked everywhere and promote them to be the "source of truth" for both gas and compute costs. In that case, I  wonder if we would want to separate two types of action costs in the profile. One for the exec cost of the current receipt and one for the send costs of newly spawned receipts. That latter is already tracked today, the former isn't.

Also, we have to consider what we can ship together with flat state. It doesn't have to be a perfect solution, yet, so we could do some hacky stuff that only works for host functions for now. But if possible we shouldn't add too much tech debt either.

Reply:
I found yet another simple, yet somewhat intermediate solution - use `#[borsh_skip]` on the `compute_usage` field in `ExecutionOutcome`. This way the serialized representation does not change and we don't need to do any migration. In essence, this is similar to adding another return value alongside `ExecutionOutcome` that is only computed and used during execution.

Longer-term, I think we should introduce this field as a first-class citizen, but in a way that will be future-proof.
Concretely, we can introduce a field storing a struct `ResourceUsage` that will contain fields like `gas_burnt`, `tokens_burnt` and `compute_usage` that are deterministically defined by the protocol and influence its' execution (as opposed to metadata which includes additional information irrelevant for execution and not enforced by the protocol). I'll file a GitHub issue to track this idea.

Reply:
That solution is pretty sweet!